### PR TITLE
Added `ArrayFlipFunctionReturnTypeExtension`, Cover non-empty-array in `array_flip`

### PR DESCRIPTION
**Pull request by staabm**
Refs https://github.com/phpstan/phpstan/issues/5327

https://3v4l.org/KUGTa

### CODE REVIEW

File: tests/PHPStan/Analyser/data/non-empty-array.php [hunk 1]
```diff
@@ -47,5 +47,7 @@ public function arrayFunctions($array, $list): void
 		assertType('array&nonEmpty', array_merge([], $array));
 		assertType('array&nonEmpty', array_merge($array, []));
 		assertType('array&nonEmpty', array_merge($array, $array));
+		
+		assertType('array&nonEmpty', array_flip($array));
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/non-empty-array.php:51]:
```suggestion
		assertType('array<(int|string)>&nonEmpty', array_flip($array));
```

**Comment by staabm** [comment]
Todo: add more test coverage

File: src/Type/Php/ArrayFlipFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+			if ($itemType !== null) {
+				$flippedArrayType = new ArrayType(
+					$itemType,
+					$keyType
+				);
+
+				if ($argType->isIterableAtLeastOnce()->yes()) {
+					$flippedArrayType = TypeCombinator::intersect($flippedArrayType, new NonEmptyArrayType());
+				}
```
Comment (staabm) [inline src/Type/Php/ArrayFlipFunctionReturnTypeExtension.php:50]: this is the actual feature I primarily want to implement with this PR.

regarding the key/type flipping I did a lot of guesswork.. I need some guidance on how this should look like :-/

File: tests/PHPStan/Analyser/data/array-flip.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ */
+function foo3($list)
+{
+	$flip = array_flip($list);
+
+	if (PHP_VERSION_ID < 80000) {
+		assertType('array|null', $flip);
+	} else {
+		assertType('array', $flip);
+	}
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/array-flip.php:27]: it seems this is not the right way to expect different types based on php version

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/array-flip.php:27]: It isn't, you need to `yield from` different files and do that `if` in the data provider.

File: src/Type/Php/ArrayFlipFunctionReturnTypeExtension.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+		$array = $functionCall->args[0]->value;
+		$argType = $scope->getType($array);
+
+		if ($argType->isArray()->yes()) {
+			$keyType = $argType->getIterableKeyType();
+			$itemType = $argType->getIterableValueType();
+
+			// make sure items, which get turned into keys contain only valid types
+			$itemType = $this->sanitizeKeyTypes($itemType);
```
Comment (ondrejmirtes) [inline src/Type/Php/ArrayFlipFunctionReturnTypeExtension.php:40]: `ArrayType::castToArrayKeyType()`

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Analyser/data/array-flip-php8.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+
+/**
+ * @param array<1|2|3, string> $list
+ */
+function foo5($array)
+{
+	$flip = array_flip($array);
+	assertType('array<string, 1|2|3>', $flip);
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/array-flip-php8.php:42]: not sure what needs to be done to get this case working

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/array-flip-php8.php:42]: Hi, you have the same debugging tools at your disposal as I have. So please look into it.

**Comment by staabm** [comment]
turns out I had a typo in my tests. should be good to go.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**